pkg/manifest/manifest.go

```diff
@@ -756,21 +792,43 @@ func UnmarshalManifest(cxt *context.Context, manifestData []byte) (*Manifest, er
 
 // SetDefaults updates the manifest with default values where not populated
 func (m *Manifest) SetDefaults() error {
-	return m.SetInvocationImageFromBundleTag(m.BundleTag)
+	return m.SetInvocationImageAndBundleTag("")
 }
 
-// SetInvocationImageFromBundleTag sets the invocation image name on the manifest
-// per the provided bundle tag, also updating the manifest BundleTag value
-// if it initially lacks a Docker tag
-func (m *Manifest) SetInvocationImageFromBundleTag(bundleTag string) error {
-	bundleRef, err := reference.ParseNormalizedNamed(bundleTag)
+// SetInvocationImageAndBundleTag sets the invocation image name and the
+// bundle tag on the manifest per the provided reference or via the
+// registry or repository values on the manifest.
+func (m *Manifest) SetInvocationImageAndBundleTag(ref string) error {
+	// Set bundle tag to repo value (combination of registry and name) by default
+	// Note: the bundle version will be appended as the Docker tag below
+	if ref != "" {
+		m.BundleTag = ref
+	} else if m.Registry != "" && m.BundleTag == "" {
```

Comment:
This if/else tree here has two concerns mixed up that makes it a bit hard to follow.

1. Composing reference from registry and bundle name when it's not set.
1. Figuring out if it should use tag or reference.
1. Validating the manifest (when everything is empty).

I'm a bit unsure what would happen if both tag and reference are set which is another validation. Maybe move all the validation into `Manifest.Validate`, default reference to tag when reference is empty, and then have a bit of code here that just deals with default the full reference using registry and name if needed.

I think that would make this easier to maintain and deprecate tag when the time comes.

Reply:
Ok, moved some logic into Validate (or, rather, a helper `validateMetadata` func).  PTAL when convenient.  Still open to suggestions for re-org/simplification.